### PR TITLE
VAULT-26466: audit - include correlation ID headers by default

### PR DESCRIPTION
**Pull request by peteski22**
## Summary 

This PR changes the behavior of Vault such that when an audit device attempts to log an incoming request which contains the following headers:

* `Correlation-ID`
* `X-Correlation-ID`

The headers, by default will appear in the audit log **without** being HMAC'ed. 

## Docs 

Docs PR: https://github.com/hashicorp/vault/pull/26778 (covers how this feature works and how to HMAC the values if required).

## Manual testing

using `/sys/config/auditing/request-headers/` endpoint:

### **No** setting for `x-correlation-id` **or** `correlation-id` headers

| Test                                           | Expected                                                         | Result |
| ----------------------------  | ------------------------------------------ | ------- |
| other headers in request        | nothing shown                                                | ✅       |
| `correlation-id` in request     | `correlation-id` shown in audit log `raw`    | ✅       |
| `x-correlation-id` in request | `x-correlation-id` shown in audit log `raw` | ✅       |

### Setting for `x-correlation-id` **and** `correlation-id` headers (no `HMAC` setting - defaults to `false`)

| Test                                           | Expected                                                         | Result |
| ----------------------------  | ------------------------------------------ | ------- |
| other headers in request        | nothing shown                                                | ✅       |
| `correlation-id` in request     | `correlation-id` shown in audit log `raw`    | ✅       |
| `x-correlation-id` in request | `x-correlation-id` shown in audit log `raw` | ✅       |

### Setting for `x-correlation-id` **and** `correlation-id` headers (`HMAC` setting: `true`)

| Test                                           | Expected                                                               | Result |
| ----------------------------  | ---------------------------------------------- | ------- |
| other headers in request        | nothing shown                                                       | ✅       |
| `correlation-id` in request     | `correlation-id` shown in audit log HMAC'ed    | ✅       |
| `x-correlation-id` in request | `x-correlation-id` shown in audit log HMAC'ed | ✅       |

### Setting for `x-correlation-id` **and** `correlation-id` headers (`HMAC` setting: `false`)

| Test                                           | Expected                                                         | Result |
| ----------------------------  | ------------------------------------------ | ------- |
| other headers in request        | nothing shown                                                | ✅       |
| `correlation-id` in request     | `correlation-id` shown in audit log `raw`    | ✅       |
| `x-correlation-id` in request | `x-correlation-id` shown in audit log `raw` | ✅       |

Example:

```shell
curl -H "X-Correlation-ID: 123-i-am-a-correlation-id-abc" -H "X-Vault-Token: hvs.du7qkklexGO3A4Ya1LjfLjJ9" http://127.0.0.1:8200/v1/sys/config/auditing/request-headers/x-correlation-id --data '{ "hmac": false }'
```

image

### CODE REVIEW

File: changelog/26777.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+```release-note:change
+audit: breaking change - Vault now allows audit logs to contain 'correlation-id' and 'x-correlation-id' headers when they
```
Comment (ccapurso) [inline changelog/26777.txt:2]: ❓ What is the user impact given the breaking change?

File: audit/headers.go [hunk 2]
```diff
@@ -163,6 +163,21 @@ func (a *HeadersConfig) Remove(ctx context.Context, header string) error {
 	return nil
 }
 
+// DefaultHeaders can be used to retrieve the set of default headers that will be
+// added to HeadersConfig in order to allow them to appear in audit logs in a raw
+// format. If the Vault Operator adds their own setting for any of the defaults,
+// their setting will be honored.
+func (a *HeadersConfig) DefaultHeaders() map[string]*HeaderSettings {
+	// Support deprecated 'x-' prefix (https://datatracker.ietf.org/doc/html/rfc6648)
+	const CorrelationID = "correlation-id"
```
[the remark this answers is not in the document]
Comment (peteski22) [reply audit/headers.go:172]: Yeah we definitely could, I think I moved it here so it was close to the one place it was being used for now, but whatever folks think works.

File: changelog/26777.txt (hunk 1 above)
Comment (peteski22) [reply changelog/26777.txt:2]: There's a corresponding docs PR that tries to call out the change: https://github.com/hashicorp/vault/pull/26778.

The Vercel deployment currently lives here: https://vault-n5poe4e19-hashicorp.vercel.app/vault/docs/audit#audit-request-headers but could change if there are any more pushes to that branch.

File: audit/headers.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ccapurso) [reply audit/headers.go]: Yeah, didn't know if we needed it anywhere else but this makes sense!

File: changelog/26777.txt (hunk 1 above)
Comment (ccapurso) [reply changelog/26777.txt:2]: Got it, thanks for the clarification!

**Review by ccapurso — APPROVED** [review]

**Pull request merged**